daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala

```diff
@@ -522,4 +558,84 @@ object Engine {
     StableConfig.copy(allowedLanguageVersions = LanguageVersion.DevVersions)
   )
 
+  private[engine] case class DisclosureTable(
+      allDisclosures: ImmArray[DisclosedContract],
+      contractIdByKey: Map[crypto.Hash, ContractId],
+      contractById: Map[ContractId, VersionedContractInstance],
+  )
+
+  private[engine] def buildDiscTable(
+      machine: Machine,
+      disclosures: ImmArray[speedy.DisclosedContract],
+  ): Result[DisclosureTable] = {
+    val acc = disclosures.foldLeft(
+      (
+        Map.empty[ContractId, VersionedContractInstance],
+        Map.empty[crypto.Hash, ContractId],
+        List.empty[DisclosedContract],
+        Option.empty[Error.Preprocessing.BadDisclosedContract],
+      )
+    ) { case ((m1, m2, ds, errorM), d) =>
+      errorM match {
+        case None =>
+          val arg = machine.normValue(d.templateId, d.argument)
+          val coid = machine.normValue(d.templateId, d.contractId)
+          val version = machine.tmplId2TxVersion(d.templateId)
+
+          // check for well typed contract argument
+          (coid, arg) match {
+            case (ValueContractId(coid), r @ ValueRecord(Some(d.templateId), _)) =>
+              // check for duplicate contract ids
+              m1.get(coid) match {
+                case Some(_) =>
+                  (
+                    m1,
+                    m2,
+                    ds,
+                    Some(Error.Preprocessing.BadDisclosedContract("Duplicate disclosed contract.")),
```

Comment:
@remyhaemmerle-da @robin-da how do you feel about using an exception for this? This seems harder to follow than just having an early exit. Alternatively a monadic fold I guess but not sure that’s a great idea in Scala.

Reply:
I am completely fine to throw an exception here.

Reply:
changed.